### PR TITLE
Percentage MGDeadlines

### PR DESCRIPTION
**Pull request by Hugo-Hontans**
Add verification on MG deadlines percentage

### CODE REVIEW

File: libs/material/src/lib/delivery/forms/informations.form.ts [hunk 1]
```diff
@@ -17,13 +17,13 @@ function createStepsFormList() {
...
     label: new FormControl(deadline.label),
-    percentage: new FormControl(deadline.percentage),
+    percentage: new FormControl(deadline.percentage, [Validators.required, validPercentage]),
     date: new FormControl(deadline.date)
   });
 }
 
 function createDeadlinesFormList() {
-  return FormList.factory([], createDeadlineFormGroup);
+  return FormList.factory([], createDeadlineFormGroup, [Validators.required, validPercentages]);
```
Comment (phpgeekfr) [inline libs/material/src/lib/delivery/forms/informations.form.ts]: `Validators.required` is not usefull here

File: libs/utils/src/lib/form/matchers.ts [hunk 2]
```diff
@@ -16,6 +16,26 @@ export class RepeatPasswordStateMatcher implements ErrorStateMatcher {
...
 }
 
+/** Checks if the sum of all percentages of FormArray does not exceed 100%  */
+export class PercentageStateMatcher implements ErrorStateMatcher {
+  private controlName: string;
+  constructor(controlName: string) {
+    this.controlName = controlName;
+  }
+
+  isErrorState(control: FormControl | null): boolean {
```
Comment (phpgeekfr) [inline libs/utils/src/lib/form/matchers.ts]: Here is another matcher, not the best though:
 ``` 
isErrorState(control: FormControl | null): boolean {
    let sum = 0;
    const controlsState = [];
    const array = control.parent.parent as FormArray;
    array.controls.forEach(group => {
      const value = group.get(this.controlName).value;
      sum += value;
      controlsState.push(group.get(this.controlName).invalid)
    });
    return (control && (sum > 100 || controlsState.some(i => i === true)));
  }
}
```

File: libs/material/src/lib/delivery/components/delivery-informations-deadlines-form/delivery-informations-deadlines-form.component.html [hunk 3]
```diff
@@ -29,16 +29,22 @@ <h4>Minimum guarantee payment deadlines</h4>
         <mat-form-field appearance="outline">
           <mat-label>Label</mat-label>
           <input type="text" formControlName="label" matInput required />
-          <mat-error *ngIf="deadline.get('label').hasError('required')">
+          <mat-error *ngIf="deadline.get('label').hasError('required') && deadline.get('label').touched">
```
Comment (phpgeekfr) [inline libs/material/src/lib/delivery/components/delivery-informations-deadlines-form/delivery-informations-deadlines-form.component.html]: testing if `deadline.get('label').touched` should not be necessary but ok ..

File: libs/material/src/lib/delivery/components/delivery-informations-deadlines-form/delivery-informations-deadlines-form.component.html [hunk 4]
```diff
@@ -29,16 +29,22 @@ <h4>Minimum guarantee payment deadlines</h4>
...
             Label is <strong>required</strong>.
           </mat-error>
         </mat-form-field>
         <mat-form-field appearance="outline">
           <mat-label>Percentage</mat-label>
-          <input type="number" min="0" max="100" formControlName="percentage" matInput required />
-          <mat-error *ngIf="deadline.get('percentage').hasError('required')">
+          <input [errorStateMatcher]="percentageMatcher" type="number" min="0" max="100" formControlName="percentage" matInput required />
+          <mat-error *ngIf="deadline.get('percentage').hasError('required') && deadline.get('percentage').touched">
```
Comment (phpgeekfr) [inline libs/material/src/lib/delivery/components/delivery-informations-deadlines-form/delivery-informations-deadlines-form.component.html]: same as above

File: libs/material/src/lib/delivery/components/delivery-informations-deadlines-form/delivery-informations-deadlines-form.component.html [hunk 5]
```diff
@@ -29,16 +29,22 @@ <h4>Minimum guarantee payment deadlines</h4>
...
             Percentage is <strong>required</strong>.
           </mat-error>
+          <mat-error *ngIf="mgDeadlines.hasError('percentageNotMatching')">
+            Percentage do not <strong>match</strong>.
```
Comment (phpgeekfr) [inline libs/material/src/lib/delivery/components/delivery-informations-deadlines-form/delivery-informations-deadlines-form.component.html]: Something more understandable ? like computed percentages cannot by above 100 or something like that

**Comment by jenna1k** [comment]
#782

**Pull request closed without merging**